bookkeeper-server/src/main/java/org/apache/bookkeeper/replication/Auditor.java

```diff
@@ -1197,6 +1197,8 @@ void checkAllLedgers() throws BKException, IOException, InterruptedException, Ke
 
             final CompletableFuture<Void> processFuture = new CompletableFuture<>();
 
+            Collection<BookieId> unavailableBookies = localAdmin.getUnavailableBookies();
```

Comment:
IIRC, checkAllLedger() potentially is a lengthy process.

The bookies in the cluster can go up or down while it is running; marking a bookie as unavailable at the beginning might be too aggressive.
Imagine this scenario:
- rack-aware data placement; 1 mil of ledger to check
- some kind of maintenance starts in the cluster (os update, bookie upgrade etc) and it is done on per-rack basis
- rack 1 is down
- check ledger starts and puts all bookies from rack 1 into the list of unavailable bookies
- after ledger 1000 is checked, the rack 1 is up, rack 2 is down
- check ledgers assumes that rack 1 bookies are down, cannot talk to the rack 2 bookies

Using BookieWatcher (and extending it) to check for the bookie availability as-needed sounds like a better approach compared to doing it once in the beginning.

Reply:
Thanks for the feedback. I hadn't realized this task was long running, and your suggestion makes sense. I'll take a look at updating this PR.

Reply:
I took your comments into consideration for the latest commit. I think it's much cleaner.